x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -102,15 +103,78 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
-  async createOrUpdateWriteTarget({ namespace }: { namespace?: string }) {
+  async rolloverAliasIfNeeded({ namespace }: { namespace?: string }) {
+    const clusterClient = await this.getClusterClient();
+    const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    let simulatedRollover: IndicesRolloverResponse;
+    try {
+      ({ body: simulatedRollover } = await clusterClient.indices.rollover({
+        alias,
+        dry_run: true,
+      }));
```

Comment:
I definitely agree that there's room for improvement on the handling of ready signals here. To me, the biggest source of confusion/possible future bugs is that the solution plugins must define a ready signal as part of step 2 (Bootstrapping specific to a security solution log) and that ready signal is then passed in to the RuleDataClient. So steps 1 and 3 are handled by the rule registry infrastructure, but solution devs still have to have fairly detailed knowledge of how the bootstrap process works to understand the requirements around "ready signal" they must implement as part of step 2.

To solve this, I envision adding a function to the `RuleDataPluginService` that would look something like `async bootstrapSolutionTemplates(logName, componentTemplates[], indexTemplate): Promise<RuleDataClient>`. This function would be responsible for executing the tasks in step 2 and then constructing a RuleDataClient to read and write from indices based on those templates. Then in the solution plugin setup we would call `const ruleDataClientPromise = ruleDataService.bootstrapSolutionTemplates(...)`, purposely not `await`ing the call to not block during setup. However, we can pass the `ruleDataClientPromise` to rule types that need it and they can `await` the promise inside the executors ensuring that `bootstrapSolutionTemplates` finishes before they use the RuleDataClient.

I think with this approach, we'd still need the ready signal internal to the RuleDataPluginService to ensure that calls to `bootstrapSolutionTemplates` don't execute until after `RuleDataPluginService.init()` completes. But, since the RuleDataClient would come from `bootstrapSolutionTemplates`, we wouldn't need a ready signal inside the RuleDataClient at all - as soon as the RuleDataClient is created by `bootstrapSolutionTemplates` it's ready to be used.

Thoughts? Do you think this would sufficiently encapsulate and simplify the bootstrap process?

Reply:
Also I'm hoping we can defer the actual implementation of this encapsulation to a follow up PR

Reply:
> Thoughts? Do you think this would sufficiently encapsulate and simplify the bootstrap process?

Yeah, this could be an option (if you really don't like the declarative way of resolving log definitions :P ). Injecting a hook. Btw, `bootstrapSolutionTemplates()` could be sync and return RuleDataClient right away. It would call the bootstrapping internally in parallel and make sure the ready signal is implemented properly.

One thing I'd keep in mind when implementing this is https://github.com/elastic/kibana/pull/102586#discussion_r656508682 - how do we make sure that inside the hook a developer doesn't have to reference common templates manually.

> Also I'm hoping we can defer the actual implementation of this encapsulation to a follow up PR

Yeah, of course, let's merge this PR sooner! 🚢

Reply:
I was thinking of `bootstrapSolutionTemplates` as a way to specify an index mapping in a declarative way rather than as a hook. Can we pair on it soon and discuss? I might be misunderstanding what declarative is in this context.